src/Store/Features.re

```diff
@@ -10,6 +10,16 @@ module Internal = {
     |> Isolinear.Effect.map(msg => Actions.Notification(msg));
   };
 
+  /* TODO: I always use the fileType from Core.Buffer.getFileType, which is bad!
+   *
+   *       This should now take into account the full filename, or the extension.
+   *       Should be future proof enough to do "ext or filename or pattern match or
+   *       first line or default."
+   *
+   *       I think this actually might be causing the current issues with .bashrc etc
+   *       now as well. .bashrc should work, but I expect the filetype is empty,
+   *       and this doesn't use the extension (which would work).
+   */
```

Comment:
This is the main open question really.

With this PR we have three ways of getting a scope back:

 - File Type
 - Extension (already here but not used as far as I can see?)
 - Filename

So this should be extracted into a suitable "try a, try b, try c" etc, but ordering there obviously matters... In say a `CMakelists.txt` from the extension its a `txt` file, but from the filename its a `cmake` file. Gets worse later with the two regex options which will be more expensive to run, so I'm a bit confused about the best ordering.

Reply:
I guess we could even use the first line, too!

We actually are setting the buffer file type here: https://github.com/onivim/oni2/blob/d9efc5b08ba63caa0bcd15f0c7e8d569d5a034de/src/Store/VimStoreConnector.re#L297 - we could potentially revise that logic, to use our new filetype resolution logic here - so that we can trust the filetype.

Reply:
1) We could update our `getLanguageFromBuffer` method in LanguageInfo to take into account the extension, filename, or even the first line (since we could check the first line with the `Buffer.t` / `Buffer.getLine`)

2) It looks like the precendence order is:
- Extensions
- Exact file name match
- First line match

https://github.com/microsoft/vscode/blob/4c99db1a05687a67c1d214361946b18e5f65e0cf/src/vs/base/common/mime.ts#L136

For now - I think we could apply that logic. Given a buffer, we'll run each of those rules, and see if there is a match (first, for extension, then for file name, then for first line) - if so, we'll apply that filetype.

Reply:
Based on this: https://github.com/microsoft/vscode/blob/4c99db1a05687a67c1d214361946b18e5f65e0cf/src/vs/base/common/mime.ts#L194-L207

I've gone with the following order, where the first of these to match is used:

 - Oni2 filename Regex (I don't really know what this is...I can't seem to see it used anywhere?)
 - Exact file name match
 - Pattern file name match (i.e. a file name regex)
 - Extension match
 - First line match

That seems to work for the more odd cases like `CMakeLists.txt` or `tsconfig.json` (which is a JSON file, defined as a `jsonc` file in the TS `package.json`).